// kepler.gl/dist/components/side-panel/interaction-panel/tooltip-config.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _reactIntl = require("react-intl");

var _styledComponents2 = require("../../common/styled-components");

var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));

var _datasetTag = _interopRequireDefault(require("../common/dataset-tag"));

var _tooltipChicklet = _interopRequireDefault(require("./tooltip-config/tooltip-chicklet"));

var _switch = _interopRequireDefault(require("../../common/switch"));

var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));

var _tooltip = require("../../../constants/tooltip");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  display: flex;\n  font-size: ", ";\n  justify-content: space-between;\n  line-height: 11px;\n  margin-bottom: 8px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: inherit;\n  padding: 0;\n\n  .button.clear-all {\n    background: transparent;\n    color: ", ";\n    margin: 0 0 0 8px;\n    padding: 0;\n\n    &:hover {\n      color: ", ";\n    }\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  .item-selector > div > div {\n    overflow: visible;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

TooltipConfigFactory.deps = [_datasetTag["default"]];

var TooltipConfigWrapper = _styledComponents["default"].div(_templateObject());

var ButtonWrapper = _styledComponents["default"].div(_templateObject2(), function (props) {
  return props.theme.subtextColor;
}, function (props) {
  return props.theme.textColor;
});

var CompareSwitchWrapper = _styledComponents["default"].div(_templateObject3(), function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.inputFontSize;
});

function TooltipConfigFactory(DatasetTag) {
  var TooltipConfig =
  /*#__PURE__*/
  function (_Component) {
    (0, _inherits2["default"])(TooltipConfig, _Component);

    function TooltipConfig() {
      (0, _classCallCheck2["default"])(this, TooltipConfig);
      return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(TooltipConfig).apply(this, arguments));
    }

    (0, _createClass2["default"])(TooltipConfig, [{
      key: "render",
      value: function render() {
        var _this$props = this.props,
            config = _this$props.config,
            datasets = _this$props.datasets,
            _onChange = _this$props.onChange,
            intl = _this$props.intl;
        return _react["default"].createElement(TooltipConfigWrapper, null, Object.keys(config.fieldsToShow).map(function (dataId) {
          return _react["default"].createElement(_styledComponents2.SidePanelSection, {
            key: dataId
          }, _react["default"].createElement(_styledComponents2.SBFlexboxNoMargin, null, _react["default"].createElement(DatasetTag, {
            dataset: datasets[dataId]
          }), Boolean(config.fieldsToShow[dataId].length) && _react["default"].createElement(ButtonWrapper, null, _react["default"].createElement(_styledComponents2.Button, {
            className: "clear-all",
            onClick: function onClick() {
              var newConfig = _objectSpread({}, config, {
                fieldsToShow: _objectSpread({}, config.fieldsToShow, (0, _defineProperty2["default"])({}, dataId, []))
              });

              _onChange(newConfig);
            },
            width: "48px",
            secondary: true
          }, _react["default"].createElement(_reactIntl.FormattedMessage, {
            id: "fieldSelector.clearAll"
          })))), _react["default"].createElement(_fieldSelector["default"], {
            fields: datasets[dataId].fields,
            value: config.fieldsToShow[dataId],
            onSelect: function onSelect(selected) {
              var newConfig = _objectSpread({}, config, {
                fieldsToShow: _objectSpread({}, config.fieldsToShow, (0, _defineProperty2["default"])({}, dataId, selected.map(function (f) {
                  return config.fieldsToShow[dataId].find(function (tooltipField) {
                    return tooltipField.name === f.name;
                  }) || {
                    name: f.name,
                    // default initial tooltip is null
                    format: null
                  };
                })))
              });

              _onChange(newConfig);
            },
            closeOnSelect: false,
            multiSelect: true,
            inputTheme: "secondary",
            CustomChickletComponent: (0, _tooltipChicklet["default"])(dataId, config, _onChange, datasets[dataId].fields)
          }));
        }), _react["default"].createElement(CompareSwitchWrapper, null, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: "compare.modeLabel"
        }), _react["default"].createElement(_switch["default"], {
          checked: config.compareMode,
          id: "compare-mode-toggle",
          onChange: function onChange() {
            var newConfig = _objectSpread({}, config, {
              compareMode: !config.compareMode
            });

            _onChange(newConfig);
          },
          secondary: true
        })), _react["default"].createElement(_styledComponents2.SidePanelSection, null, _react["default"].createElement(_styledComponents2.PanelLabel, null, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: "compare.typeLabel"
        })), _react["default"].createElement(_itemSelector["default"], {
          disabled: !config.compareMode,
          displayOption: function displayOption(d) {
            return intl.formatMessage({
              id: "compare.types.".concat(d)
            });
          },
          selectedItems: config.compareType,
          options: Object.values(_tooltip.COMPARE_TYPES),
          multiSelect: false,
          searchable: false,
          inputTheme: 'secondary',
          getOptionValue: function getOptionValue(d) {
            return d;
          },
          onChange: function onChange(option) {
            var newConfig = _objectSpread({}, config, {
              compareType: option
            });

            _onChange(newConfig);
          }
        })));
      }
    }]);
    return TooltipConfig;
  }(_react.Component);

  return (0, _reactIntl.injectIntl)(TooltipConfig);
}

var _default = TooltipConfigFactory;
exports["default"] = _default;
