// kepler.gl/dist/components/common/time-slider-marker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Scale = require("d3-scale");

var _d3Selection = require("d3-selection");

var _d3Axis = require("d3-axis");

var _reselect = require("reselect");

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  pointer-events: none;\n  position: absolute;\n\n  .axis text {\n    font-size: 9px;\n    fill: ", ";\n  }\n\n  .axis line,\n  .axis path {\n    fill: none;\n    stroke: ", ";\n    shape-rendering: crispEdges;\n    stroke-width: 2;\n  }\n\n  .axis .domain {\n    display: none;\n  }\n\n  .value {\n    fill: ", ";\n    font-size: 10px;\n\n    &.start {\n      text-anchor: start;\n    }\n\n    &.end {\n      text-anchor: end;\n    }\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var MIN_TICK_WIDTH = 50;

var TimeSliderContainer = _styledComponents["default"].svg(_templateObject(), function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.sliderBarBgd;
}, function (props) {
  return props.theme.textColor;
});

var height = 30;

function TimeSliderMarkerFactory() {
  var TimeSliderMarker =
  /*#__PURE__*/
  function (_Component) {
    (0, _inherits2["default"])(TimeSliderMarker, _Component);

    function TimeSliderMarker() {
      var _getPrototypeOf2;

      var _this;

      (0, _classCallCheck2["default"])(this, TimeSliderMarker);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(TimeSliderMarker)).call.apply(_getPrototypeOf2, [this].concat(args)));
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "xAxis", (0, _react.createRef)());
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "domainSelector", function (props) {
        return props.domain;
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "widthSelector", function (props) {
        return props.width;
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "scaleSelector", (0, _reselect.createSelector)(_this.domainSelector, _this.widthSelector, function (domain, width) {
        return Array.isArray(domain) ? (0, _d3Scale.scaleUtc)().domain(domain).range([0, width]) : null;
      }));
      return _this;
    }

    (0, _createClass2["default"])(TimeSliderMarker, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        this._updateAxis(this.scaleSelector(this.props));
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate(prevProps) {
        if (this.scaleSelector(this.props) !== this.scaleSelector(prevProps)) {
          this._updateAxis(this.scaleSelector(this.props));
        }
      }
    }, {
      key: "_updateAxis",
      value: function _updateAxis(scale) {
        if (!scale) {
          return;
        } // TODO: pass in ticks if interval is defined


        var ticks = Math.floor(this.props.width / MIN_TICK_WIDTH);
        var xAxis = (0, _d3Axis.axisBottom)(scale).ticks(ticks).tickSize(8).tickPadding(6);
        (0, _d3Selection.select)(this.xAxis.current).call(xAxis);
      }
    }, {
      key: "render",
      value: function render() {
        return _react["default"].createElement(TimeSliderContainer, {
          className: "time-slider-marker",
          width: this.props.width,
          height: height
        }, _react["default"].createElement("g", {
          className: "x axis",
          ref: this.xAxis,
          transform: "translate(0, 0)"
        }));
      }
    }]);
    return TimeSliderMarker;
  }(_react.Component);

  (0, _defineProperty2["default"])(TimeSliderMarker, "propTypes", {
    domain: _propTypes["default"].arrayOf(_propTypes["default"].any).isRequired,
    width: _propTypes["default"].number.isRequired
  });
  return TimeSliderMarker;
}

var _default = TimeSliderMarkerFactory;
exports["default"] = _default;
