// kepler.gl/dist/components/modals/export-image-modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _imagePreview = _interopRequireDefault(require("../common/image-preview"));

var _defaultSettings = require("../../constants/default-settings");

var _styledComponents2 = require("../common/styled-components");

var _switch = _interopRequireDefault(require("../common/switch"));

var _reactIntl = require("react-intl");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  justify-content: space-around;\n  width: 250px;\n\n  .image-option-section {\n    .image-option-section-title {\n      font-weight: 500;\n      font-size: 14px;\n    }\n  }\n\n  .button-list {\n    display: flex;\n    flex-direction: row;\n    padding: 8px 0px;\n  }\n\n  input {\n    margin-right: 8px;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var ImageOptionList = _styledComponents["default"].div(_templateObject());

var ExportImageModalFactory = function ExportImageModalFactory() {
  var ExportImageModal =
  /*#__PURE__*/
  function (_Component) {
    (0, _inherits2["default"])(ExportImageModal, _Component);

    function ExportImageModal() {
      (0, _classCallCheck2["default"])(this, ExportImageModal);
      return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ExportImageModal).apply(this, arguments));
    }

    (0, _createClass2["default"])(ExportImageModal, [{
      key: "componentDidUpdate",
      value: function componentDidUpdate() {
        this._updateMapDim();
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        this.props.onUpdateImageSetting({
          exporting: false
        });
      }
    }, {
      key: "_updateMapDim",
      value: function _updateMapDim() {
        var _this$props = this.props,
            exportImage = _this$props.exportImage,
            mapH = _this$props.mapH,
            mapW = _this$props.mapW;

        if (mapH !== exportImage.mapH || mapW !== exportImage.mapW) {
          this.props.onUpdateImageSetting({
            mapH: mapH,
            mapW: mapW,
            legend: false
          });
        }
      }
    }, {
      key: "render",
      value: function render() {
        var _this$props2 = this.props,
            exportImage = _this$props2.exportImage,
            onUpdateImageSetting = _this$props2.onUpdateImageSetting,
            intl = _this$props2.intl;
        var legend = exportImage.legend,
            ratio = exportImage.ratio,
            resolution = exportImage.resolution;
        return _react["default"].createElement(_styledComponents2.StyledModalContent, {
          className: "export-image-modal"
        }, _react["default"].createElement(ImageOptionList, null, _react["default"].createElement("div", {
          className: "image-option-section"
        }, _react["default"].createElement("div", {
          className: "image-option-section-title"
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: 'modal.exportImage.ratioTitle'
        })), _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: 'modal.exportImage.ratioDescription'
        }), _react["default"].createElement("div", {
          className: "button-list"
        }, _defaultSettings.EXPORT_IMG_RATIO_OPTIONS.filter(function (op) {
          return !op.hidden;
        }).map(function (op) {
          return _react["default"].createElement(_styledComponents2.SelectionButton, {
            key: op.id,
            selected: ratio === op.id,
            onClick: function onClick() {
              return onUpdateImageSetting({
                ratio: op.id
              });
            }
          }, _react["default"].createElement(_reactIntl.FormattedMessage, {
            id: op.label
          }));
        }))), _react["default"].createElement("div", {
          className: "image-option-section"
        }, _react["default"].createElement("div", {
          className: "image-option-section-title"
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: 'modal.exportImage.resolutionTitle'
        })), _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: 'modal.exportImage.resolutionDescription'
        }), _react["default"].createElement("div", {
          className: "button-list"
        }, _defaultSettings.EXPORT_IMG_RESOLUTION_OPTIONS.map(function (op) {
          return _react["default"].createElement(_styledComponents2.SelectionButton, {
            key: op.id,
            selected: resolution === op.id,
            onClick: function onClick() {
              return op.available && onUpdateImageSetting({
                resolution: op.id
              });
            }
          }, op.label);
        }))), _react["default"].createElement("div", {
          className: "image-option-section"
        }, _react["default"].createElement("div", {
          className: "image-option-section-title"
        }, _react["default"].createElement(_reactIntl.FormattedMessage, {
          id: 'modal.exportImage.mapLegendTitle'
        })), _react["default"].createElement(_switch["default"], {
          type: "checkbox",
          id: "add-map-legend",
          checked: legend,
          label: intl.formatMessage({
            id: 'modal.exportImage.mapLegendAdd'
          }),
          onChange: function onChange() {
            return onUpdateImageSetting({
              legend: !legend
            });
          }
        }))), _react["default"].createElement(_imagePreview["default"], {
          exportImage: exportImage
        }));
      }
    }]);
    return ExportImageModal;
  }(_react.Component);

  (0, _defineProperty2["default"])(ExportImageModal, "propTypes", {
    mapW: _propTypes["default"].number.isRequired,
    mapH: _propTypes["default"].number.isRequired,
    exportImage: _propTypes["default"].object.isRequired,
    // callbacks
    onUpdateImageSetting: _propTypes["default"].func.isRequired
  });
  return (0, _reactIntl.injectIntl)(ExportImageModal);
};

var _default = ExportImageModalFactory;
exports["default"] = _default;
