// kepler.gl/dist/components/common/animation-control/animation-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _moment = _interopRequireDefault(require("moment"));

var _slider = _interopRequireDefault(require("../slider/slider"));

var _styledComponents2 = require("../styled-components");

var _speedControl = _interopRequireDefault(require("./speed-control"));

var _playbackControls = _interopRequireDefault(require("./playback-controls"));

var _floatingTimeDisplay = _interopRequireDefault(require("./floating-time-display"));

var _animationController = _interopRequireDefault(require("./animation-controller"));

var _dataUtils = require("../../../utils/data-utils");

var _constants = require("../../../constants");

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  font-weight: 400;\n  font-size: 10px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  display: flex;\n  align-items: center;\n  height: 32px;\n\n  .animation-control__speed-control {\n    margin-right: -10px;\n\n    .animation-control__speed-slider {\n      right: calc(0% - 10px);\n    }\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  position: relative;\n  flex-grow: 1;\n  margin-right: 24px;\n  margin-left: 24px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var SliderWrapper = _styledComponents["default"].div(_templateObject());

var AnimationWidgetInner = _styledComponents["default"].div(_templateObject2());

var StyledDomain = _styledComponents["default"].div(_templateObject3(), function (props) {
  return props.theme.titleTextColor;
});

var BUTTON_HEIGHT = '18px';
AnimationControlFactory.deps = [_speedControl["default"], _playbackControls["default"], _floatingTimeDisplay["default"], _animationController["default"]];

function AnimationControlFactory(SpeedControl, AnimationPlaybacks, FloatingTimeDisplay, AnimationController) {
  var AnimationControl =
  /*#__PURE__*/
  function (_React$PureComponent) {
    (0, _inherits2["default"])(AnimationControl, _React$PureComponent);

    function AnimationControl() {
      var _getPrototypeOf2;

      var _this;

      (0, _classCallCheck2["default"])(this, AnimationControl);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(AnimationControl)).call.apply(_getPrototypeOf2, [this].concat(args)));
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
        showSpeedControl: false
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onSlider1Change", function (val) {
        var domain = _this.props.animationConfig.domain;

        if (Array.isArray(_this.props.timeSteps)) {
          _this.props.updateAnimationTime((0, _dataUtils.snapToMarks)(val, _this.props.timeSteps)); // TODO: merge slider in to avoid this step

        } else if (val >= domain[0] && val <= domain[1]) {
          _this.props.updateAnimationTime(val);
        }
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "toggleSpeedControl", function () {
        _this.setState({
          showSpeedControl: !_this.state.showSpeedControl
        });
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "onChange", function () {
        _this.toggleSpeedControl();
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_updateAnimation", function (value) {
        _this.props.updateAnimationTime(Array.isArray(value) ? value[0] : value);
      });
      return _this;
    }

    (0, _createClass2["default"])(AnimationControl, [{
      key: "render",
      value: function render() {
        var _this$props$animation = this.props.animationConfig,
            currentTime = _this$props$animation.currentTime,
            domain = _this$props$animation.domain,
            speed = _this$props$animation.speed,
            step = _this$props$animation.step,
            timeSteps = _this$props$animation.timeSteps;
        var showSpeedControl = this.state.showSpeedControl;
        var animationType = Array.isArray(timeSteps) ? _constants.ANIMATION_TYPE.interval : _constants.ANIMATION_TYPE.continuous;
        return _react["default"].createElement(_styledComponents2.BottomWidgetInner, {
          className: "bottom-widget--inner"
        }, _react["default"].createElement(AnimationWidgetInner, {
          className: "animation-widget--inner"
        }, _react["default"].createElement("div", {
          style: {
            marginLeft: '-10px'
          }
        }, _react["default"].createElement(AnimationController, {
          value: currentTime,
          domain: domain,
          speed: speed,
          updateAnimation: this._updateAnimation,
          steps: timeSteps,
          animationType: animationType
        }, function (isAnimating, start, pause, reset) {
          return _react["default"].createElement(AnimationPlaybacks, {
            className: "animation-control-playpause",
            startAnimation: start,
            isAnimating: isAnimating,
            pauseAnimation: pause,
            resetAnimation: reset,
            buttonHeight: BUTTON_HEIGHT,
            buttonStyle: "link"
          });
        })), _react["default"].createElement(StyledDomain, {
          className: "animation-control__time-domain"
        }, _react["default"].createElement("span", null, _moment["default"].utc(domain[0]).format(_constants.DEFAULT_TIME_FORMAT))), _react["default"].createElement(SliderWrapper, {
          className: "animation-control__slider"
        }, _react["default"].createElement(_slider["default"], {
          showValues: false,
          isRanged: false,
          step: step,
          minValue: domain ? domain[0] : 0,
          maxValue: domain ? domain[1] : 1,
          value1: currentTime,
          onSlider1Change: this.onSlider1Change,
          enableBarDrag: true
        })), _react["default"].createElement(StyledDomain, {
          className: "animation-control__time-domain"
        }, _react["default"].createElement("span", null, _moment["default"].utc(domain[1]).format(_constants.DEFAULT_TIME_FORMAT))), _react["default"].createElement("div", {
          className: "animation-control__speed-control"
        }, _react["default"].createElement(SpeedControl, {
          onClick: this.toggleSpeedControl,
          showSpeedControl: showSpeedControl,
          updateAnimationSpeed: this.props.updateAnimationSpeed,
          speed: speed,
          buttonHeight: BUTTON_HEIGHT
        }))), _react["default"].createElement(FloatingTimeDisplay, {
          currentTime: currentTime
        }));
      }
    }]);
    return AnimationControl;
  }(_react["default"].PureComponent);

  AnimationControl.defaultProps = {
    sliderHandleWidth: 12,
    onChange: function onChange() {}
  };
  return AnimationControl;
}

var _default = AnimationControlFactory;
exports["default"] = _default;
