// @nebula.gl/edit-modes/dist-es6/lib/edit-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseEditMode = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var BaseEditMode =
/*#__PURE__*/
function () {
  function BaseEditMode() {
    _classCallCheck(this, BaseEditMode);

    _defineProperty(this, "state", void 0);
  }

  _createClass(BaseEditMode, [{
    key: "getState",
    value: function getState() {
      return this.state;
    }
  }, {
    key: "updateState",
    value: function updateState(state) {
      var _this = this;

      var prevState = this.state;
      var changedEvents = [];

      if (!this.state || this.state.data !== state.data) {
        changedEvents.push(this.onDataChanged);
      }

      if (!this.state || this.state.modeConfig !== state.modeConfig) {
        changedEvents.push(this.onModeConfigChanged);
      }

      if (!this.state || this.state.selectedIndexes !== state.selectedIndexes) {
        changedEvents.push(this.onSelectedIndexesChanged);
      }

      if (!this.state || this.state.guides !== state.guides) {
        changedEvents.push(this.onGuidesChanged);
      }

      this.state = state;
      changedEvents.forEach(function (fn) {
        return fn.bind(_this)(prevState);
      });
    } // Overridable user interaction handlers

  }, {
    key: "handleClick",
    value: function handleClick(event) {}
  }, {
    key: "handlePointerMove",
    value: function handlePointerMove(event) {}
  }, {
    key: "handleStartDragging",
    value: function handleStartDragging(event) {}
  }, {
    key: "handleStopDragging",
    value: function handleStopDragging(event) {} // Convenience functions to handle state changes

  }, {
    key: "onDataChanged",
    value: function onDataChanged(prevState) {}
  }, {
    key: "onModeConfigChanged",
    value: function onModeConfigChanged(prevState) {}
  }, {
    key: "onSelectedIndexesChanged",
    value: function onSelectedIndexesChanged(prevState) {}
  }, {
    key: "onGuidesChanged",
    value: function onGuidesChanged(prevState) {} // Convenience functions to access state

  }, {
    key: "getData",
    value: function getData() {
      return this.state.data;
    }
  }, {
    key: "getModeConfig",
    value: function getModeConfig() {
      return this.state.modeConfig;
    }
  }, {
    key: "getSelectedIndexes",
    value: function getSelectedIndexes() {
      return this.state.selectedIndexes;
    }
  }, {
    key: "getGuides",
    value: function getGuides() {
      return this.state && this.state.guides;
    }
  }, {
    key: "getCursor",
    value: function getCursor() {
      return this.state && this.state.cursor;
    }
  }, {
    key: "onEdit",
    value: function onEdit(editAction) {
      this.state.onEdit(editAction);
    }
  }, {
    key: "onUpdateGuides",
    value: function onUpdateGuides(guides) {
      this.state.onUpdateGuides(guides);
    }
  }, {
    key: "onUpdateCursor",
    value: function onUpdateCursor(cursor) {
      this.state.onUpdateCursor(cursor);
    }
  }]);

  return BaseEditMode;
}();

exports.BaseEditMode = BaseEditMode;
