// kepler.gl/dist/components/common/data-table/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.DataTable = exports.TableSection = exports.Container = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactVirtualized = require("react-virtualized");

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _reselect = require("reselect");

var _lodash = _interopRequireDefault(require("lodash.get"));

var _lodash2 = _interopRequireDefault(require("lodash.debounce"));

var _optionDropdown = _interopRequireDefault(require("./option-dropdown"));

var _grid = _interopRequireDefault(require("./grid"));

var _button = _interopRequireDefault(require("./button"));

var _icons = require("../icons");

var _fieldToken = _interopRequireDefault(require("../field-token"));

var _dataUtils = require("../../../utils/data-utils");

var _cellSize = require("./cell-size");

var _defaultSettings = require("../../../constants/default-settings");

var _fieldToAlignRight;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  font-size: 11px;\n  flex-grow: 1;\n  color: ", ";\n  width: 100%;\n\n  .ReactVirtualized__Grid:focus,\n  .ReactVirtualized__Grid:active {\n    outline: 0;\n  }\n\n  .cell {\n    &::-webkit-scrollbar {\n      display: none;\n    }\n  }\n\n  *:focus {\n    outline: 0;\n  }\n\n  .results-table-wrapper {\n    position: relative;\n    min-height: 100%;\n    max-height: 100%;\n    display: flex;\n    flex-direction: row;\n    flex-grow: 1;\n    overflow: hidden;\n    border-top: none;\n\n    .scroll-in-ui-thread::after {\n      content: '';\n      height: 100%;\n      left: 0;\n      position: absolute;\n      pointer-events: none;\n      top: 0;\n      width: 100%;\n    }\n\n    .grid-row {\n      position: relative;\n      display: flex;\n      flex-direction: row;\n    }\n    .grid-column {\n      display: flex;\n      flex-direction: column;\n      flex: 1 1 auto;\n    }\n    .pinned-grid-container {\n      flex: 0 0 75px;\n      z-index: 10;\n      position: absolute;\n      left: 0;\n      top: 0;\n      border-right: 2px solid ", ";\n    }\n\n    .header-grid {\n      overflow: hidden !important;\n    }\n\n    .body-grid {\n      overflow: overlay !important;\n    }\n\n    .pinned-grid {\n      overflow: overlay !important;\n    }\n\n    .even-row {\n      background-color: ", ";\n    }\n    .odd-row {\n      background-color: ", ";\n    }\n    .cell,\n    .header-cell {\n      width: 100%;\n      height: 100%;\n      display: flex;\n      flex-direction: column;\n      justify-content: center;\n      align-items: flex-start;\n      text-align: center;\n      overflow: hidden;\n\n      .n-sort-idx {\n        font-size: 9px;\n      }\n    }\n    .cell {\n      border-bottom: 1px solid ", ";\n      border-right: 1px solid ", ";\n      white-space: nowrap;\n      overflow: auto;\n      padding: 0 ", "px;\n      font-size: ", "px;\n\n      .result-link {\n        text-decoration: none;\n      }\n    }\n    .cell.end-cell,\n    .header-cell.end-cell {\n      border-right: none;\n      padding-right: ", "px;\n    }\n    .cell.first-cell,\n    .header-cell.first-cell {\n      padding-left: ", "px;\n    }\n    .cell.bottom-cell {\n      border-bottom: none;\n    }\n    .cell.align-right {\n      align-items: flex-end;\n    }\n    .header-cell {\n      border-bottom: 1px solid ", ";\n      border-top: 1px solid ", ";\n      padding-top: ", "px;\n      padding-right: 0;\n      padding-bottom: ", "px;\n      padding-left: ", "px;\n      align-items: center;\n      justify-content: space-between;\n      display: flex;\n      flex-direction: row;\n      background-color: ", ";\n\n      &:hover {\n        .more {\n          color: ", ";\n        }\n      }\n      .n-sort-idx {\n        font-size: 9px;\n      }\n      .details {\n        font-weight: 500;\n        display: flex;\n        flex-direction: column;\n        justify-content: flex-start;\n        height: 100%;\n        overflow: hidden;\n        flex-grow: 1;\n        .col-name {\n          display: flex;\n          align-items: center;\n          justify-content: space-between;\n\n          .col-name__left {\n            display: flex;\n            align-items: center;\n            overflow: hidden;\n            svg {\n              margin-left: 6px;\n            }\n          }\n          .col-name__name {\n            overflow: hidden;\n            white-space: nowrap;\n          }\n          .col-name__sort {\n            cursor: pointer;\n          }\n        }\n      }\n\n      .more {\n        color: transparent;\n        margin-left: 5px;\n      }\n    }\n  }\n\n  :focus {\n    outline: none;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var defaultHeaderRowHeight = 55;
var defaultRowHeight = 32;
var overscanColumnCount = 10;
var overscanRowCount = 10;
var fieldToAlignRight = (_fieldToAlignRight = {}, (0, _defineProperty2["default"])(_fieldToAlignRight, _defaultSettings.ALL_FIELD_TYPES.integer, true), (0, _defineProperty2["default"])(_fieldToAlignRight, _defaultSettings.ALL_FIELD_TYPES.real, true), _fieldToAlignRight);

var Container = _styledComponents["default"].div(_templateObject(), function (props) {
  return props.theme.textColorLT;
}, function (props) {
  return props.theme.pinnedGridBorderColor;
}, function (props) {
  return props.theme.evenRowBackground;
}, function (props) {
  return props.theme.oddRowBackground;
}, function (props) {
  return props.theme.cellBorderColor;
}, function (props) {
  return props.theme.cellBorderColor;
}, function (props) {
  return props.theme.cellPaddingSide;
}, function (props) {
  return props.theme.cellFontSize;
}, function (props) {
  return props.theme.cellPaddingSide + props.theme.edgeCellPaddingSide;
}, function (props) {
  return props.theme.cellPaddingSide + props.theme.edgeCellPaddingSide;
}, function (props) {
  return props.theme.headerCellBorderColor;
}, function (props) {
  return props.theme.headerCellBorderColor;
}, function (props) {
  return props.theme.headerPaddingTop;
}, function (props) {
  return props.theme.headerPaddingBottom;
}, function (props) {
  return props.theme.cellPaddingSide;
}, function (props) {
  return props.theme.headerCellBackground;
}, function (props) {
  return props.theme.headerCellIconColor;
});

exports.Container = Container;
var defaultColumnWidth = 200;

var columnWidthFunction = function columnWidthFunction(columns, cellSizeCache, ghost) {
  return function (_ref) {
    var index = _ref.index;
    return (columns[index] || {}).ghost ? ghost : cellSizeCache[columns[index]] || defaultColumnWidth;
  };
};
/*
 * This is an accessor method used to generalize getting a cell from a data row
 */


var getRowCell = function getRowCell(_ref2) {
  var rows = _ref2.rows,
      columns = _ref2.columns,
      column = _ref2.column,
      colMeta = _ref2.colMeta,
      rowIndex = _ref2.rowIndex,
      sortColumn = _ref2.sortColumn,
      sortOrder = _ref2.sortOrder;
  var rowIdx = sortOrder && sortOrder.length ? (0, _lodash["default"])(sortOrder, rowIndex) : rowIndex;
  var type = colMeta[column];
  return (0, _dataUtils.parseFieldValue)((0, _lodash["default"])(rows, [rowIdx, columns.indexOf(column)], 'Err'), type);
};

var renderHeaderCell = function renderHeaderCell(columns, isPinned, props, toggleMoreOptions, moreOptionsColumn) {
  // eslint-disable-next-line react/display-name
  return function (cellInfo) {
    var _classnames;

    var columnIndex = cellInfo.columnIndex,
        key = cellInfo.key,
        style = cellInfo.style;
    var colMeta = props.colMeta,
        sortColumn = props.sortColumn,
        _sortTableColumn = props.sortTableColumn,
        unsortColumn = props.unsortColumn,
        _pinTableColumn = props.pinTableColumn,
        _copyTableColumn = props.copyTableColumn,
        dataId = props.dataId;
    var column = columns[columnIndex];
    var isGhost = column.ghost;
    var isSorted = sortColumn[column];
    var firstCell = columnIndex === 0;
    return _react["default"].createElement("div", {
      className: (0, _classnames3["default"])('header-cell', (_classnames = {}, (0, _defineProperty2["default"])(_classnames, "column-".concat(columnIndex), true), (0, _defineProperty2["default"])(_classnames, 'pinned-header-cell', isPinned), (0, _defineProperty2["default"])(_classnames, 'first-cell', firstCell), _classnames)),
      key: key,
      style: style,
      onClick: function onClick(e) {
        e.shiftKey ? _sortTableColumn(dataId, column) : null;
      },
      onDoubleClick: function onDoubleClick() {
        return _sortTableColumn(dataId, column);
      },
      title: column
    }, isGhost ? _react["default"].createElement("div", null) : _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("section", {
      className: "details"
    }, _react["default"].createElement("div", {
      className: "col-name"
    }, _react["default"].createElement("div", {
      className: "col-name__left"
    }, _react["default"].createElement("div", {
      className: "col-name__name"
    }, column), _react["default"].createElement(_button["default"], {
      className: "col-name__sort",
      onClick: function onClick() {
        return _sortTableColumn(dataId, column);
      }
    }, isSorted ? isSorted === _defaultSettings.SORT_ORDER.ASCENDING ? _react["default"].createElement(_icons.ArrowUp, {
      height: "14px"
    }) : _react["default"].createElement(_icons.ArrowDown, {
      height: "14px"
    }) : null)), _react["default"].createElement(_button["default"], {
      className: "more",
      onClick: function onClick() {
        return toggleMoreOptions(column);
      }
    }, _react["default"].createElement(_icons.VertThreeDots, {
      height: "14px"
    }))), _react["default"].createElement(_fieldToken["default"], {
      type: colMeta[column]
    })), _react["default"].createElement("section", {
      className: "options"
    }, _react["default"].createElement(_optionDropdown["default"], {
      isOpened: moreOptionsColumn === column,
      type: colMeta[column],
      column: column,
      toggleMoreOptions: toggleMoreOptions,
      sortTableColumn: function sortTableColumn(mode) {
        return _sortTableColumn(dataId, column, mode);
      },
      sortMode: sortColumn && sortColumn[column],
      pinTableColumn: function pinTableColumn() {
        return _pinTableColumn(dataId, column);
      },
      copyTableColumn: function copyTableColumn() {
        return _copyTableColumn(dataId, column);
      },
      isSorted: isSorted,
      isPinned: isPinned,
      unsortColumn: unsortColumn
    }))));
  };
};

var renderDataCell = function renderDataCell(columns, isPinned, props) {
  return function (cellInfo) {
    var _classnames2;

    var columnIndex = cellInfo.columnIndex,
        key = cellInfo.key,
        style = cellInfo.style,
        rowIndex = cellInfo.rowIndex;
    var rows = props.rows,
        colMeta = props.colMeta;
    var column = columns[columnIndex];
    var isGhost = column.ghost;
    var rowCell = isGhost ? '' : getRowCell(_objectSpread({}, props, {
      column: column,
      rowIndex: rowIndex
    }));
    var type = isGhost ? null : colMeta[column];
    var endCell = columnIndex === columns.length - 1;
    var firstCell = columnIndex === 0;
    var bottomCell = rowIndex === rows.length - 1;
    var alignRight = fieldToAlignRight[type];

    var cell = _react["default"].createElement("div", {
      className: (0, _classnames3["default"])('cell', (_classnames2 = {}, (0, _defineProperty2["default"])(_classnames2, rowIndex % 2 === 0 ? 'even-row' : 'odd-row', true), (0, _defineProperty2["default"])(_classnames2, "row-".concat(rowIndex), true), (0, _defineProperty2["default"])(_classnames2, 'pinned-cell', isPinned), (0, _defineProperty2["default"])(_classnames2, 'first-cell', firstCell), (0, _defineProperty2["default"])(_classnames2, 'end-cell', endCell), (0, _defineProperty2["default"])(_classnames2, 'bottom-cell', bottomCell), (0, _defineProperty2["default"])(_classnames2, 'align-right', alignRight), _classnames2)),
      key: key,
      style: style,
      title: isGhost ? undefined : rowCell
    }, "".concat(rowCell).concat(endCell ? '\n' : '\t'));

    return cell;
  };
};

var TableSection = function TableSection(_ref3) {
  var classList = _ref3.classList,
      isPinned = _ref3.isPinned,
      columns = _ref3.columns,
      headerGridProps = _ref3.headerGridProps,
      fixedWidth = _ref3.fixedWidth,
      _ref3$fixedHeight = _ref3.fixedHeight,
      fixedHeight = _ref3$fixedHeight === void 0 ? undefined : _ref3$fixedHeight,
      onScroll = _ref3.onScroll,
      scrollTop = _ref3.scrollTop,
      dataGridProps = _ref3.dataGridProps,
      columnWidth = _ref3.columnWidth,
      setGridRef = _ref3.setGridRef,
      headerCellRender = _ref3.headerCellRender,
      dataCellRender = _ref3.dataCellRender,
      _ref3$scrollLeft = _ref3.scrollLeft,
      scrollLeft = _ref3$scrollLeft === void 0 ? undefined : _ref3$scrollLeft;
  return _react["default"].createElement(_reactVirtualized.AutoSizer, null, function (_ref4) {
    var width = _ref4.width,
        height = _ref4.height;
    var gridDimension = {
      columnCount: columns.length,
      columnWidth: columnWidth,
      width: fixedWidth || width
    };
    var dataGridHeight = fixedHeight || height;
    return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("div", {
      className: (0, _classnames3["default"])('scroll-in-ui-thread', classList.header)
    }, _react["default"].createElement(_grid["default"], (0, _extends2["default"])({
      cellRenderer: headerCellRender
    }, headerGridProps, gridDimension, {
      scrollLeft: scrollLeft
    }))), _react["default"].createElement("div", {
      className: (0, _classnames3["default"])('scroll-in-ui-thread', classList.rows),
      style: {
        top: headerGridProps.height
      }
    }, _react["default"].createElement(_grid["default"], (0, _extends2["default"])({
      cellRenderer: dataCellRender
    }, dataGridProps, gridDimension, {
      className: isPinned ? 'pinned-grid' : 'body-grid',
      height: dataGridHeight - headerGridProps.height,
      onScroll: onScroll,
      scrollTop: scrollTop,
      setGridRef: setGridRef
    }))));
  });
};

exports.TableSection = TableSection;

var DataTable =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(DataTable, _Component);

  function DataTable() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, DataTable);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(DataTable)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      cellSizeCache: {},
      moreOptionsColumn: null
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "root", (0, _react.createRef)());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "columns", function (props) {
      return props.columns;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "pinnedColumns", function (props) {
      return props.pinnedColumns;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "unpinnedColumns", (0, _reselect.createSelector)(_this.columns, _this.pinnedColumns, function (columns, pinnedColumns) {
      return !Array.isArray(pinnedColumns) ? columns : columns.filter(function (c) {
        return !pinnedColumns.includes(c);
      });
    }));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "toggleMoreOptions", function (moreOptionsColumn) {
      return _this.setState({
        moreOptionsColumn: _this.state.moreOptionsColumn === moreOptionsColumn ? null : moreOptionsColumn
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "getCellSizeCache", function () {
      var _this$props = _this.props,
          propsCache = _this$props.cellSizeCache,
          fixedWidth = _this$props.fixedWidth,
          pinnedColumns = _this$props.pinnedColumns;

      var unpinnedColumns = _this.unpinnedColumns(_this.props);

      var width = fixedWidth ? fixedWidth : _this.root.current ? _this.root.current.clientWidth : 0; // pin column border is 2 pixel vs 1 pixel

      var adjustWidth = pinnedColumns.length ? width - 1 : width;

      var _adjustCellsToContain = (0, _cellSize.adjustCellsToContainer)(adjustWidth, propsCache, pinnedColumns, unpinnedColumns),
          cellSizeCache = _adjustCellsToContain.cellSizeCache,
          ghost = _adjustCellsToContain.ghost;

      return {
        cellSizeCache: cellSizeCache,
        ghost: ghost
      };
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "doScaleCellsToWidth", function () {
      _this.setState(_this.getCellSizeCache());
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "scaleCellsToWidth", (0, _lodash2["default"])(_this.doScaleCellsToWidth, 300));
    return _this;
  }

  (0, _createClass2["default"])(DataTable, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      window.addEventListener('resize', this.scaleCellsToWidth);
      this.scaleCellsToWidth();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.props.cellSizeCache !== prevProps.cellSizeCache || this.props.pinnedColumns !== prevProps.pinnedColumns) {
        this.scaleCellsToWidth();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.scaleCellsToWidth);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          rows = _this$props2.rows,
          pinnedColumns = _this$props2.pinnedColumns,
          _this$props2$theme = _this$props2.theme,
          theme = _this$props2$theme === void 0 ? {} : _this$props2$theme,
          fixedWidth = _this$props2.fixedWidth,
          fixedHeight = _this$props2.fixedHeight;
      var unpinnedColumns = this.unpinnedColumns(this.props);
      var _this$state = this.state,
          cellSizeCache = _this$state.cellSizeCache,
          moreOptionsColumn = _this$state.moreOptionsColumn,
          ghost = _this$state.ghost;
      var unpinnedColumnsGhost = ghost ? [].concat((0, _toConsumableArray2["default"])(unpinnedColumns), [{
        ghost: true
      }]) : unpinnedColumns;
      var pinnedColumnsWidth = pinnedColumns.reduce(function (acc, val) {
        return acc + (0, _lodash["default"])(cellSizeCache, val, 0);
      }, 0);
      var hasPinnedColumns = Boolean(pinnedColumns.length);
      var _theme$headerRowHeigh = theme.headerRowHeight,
          headerRowHeight = _theme$headerRowHeigh === void 0 ? defaultHeaderRowHeight : _theme$headerRowHeigh,
          _theme$rowHeight = theme.rowHeight,
          rowHeight = _theme$rowHeight === void 0 ? defaultRowHeight : _theme$rowHeight;
      var headerGridProps = {
        cellSizeCache: cellSizeCache,
        className: 'header-grid',
        height: headerRowHeight,
        rowCount: 1,
        rowHeight: headerRowHeight
      };
      var dataGridProps = {
        cellSizeCache: cellSizeCache,
        overscanColumnCount: overscanColumnCount,
        overscanRowCount: overscanRowCount,
        rowCount: (rows || []).length,
        rowHeight: rowHeight
      };
      return _react["default"].createElement(Container, {
        className: "data-table-container",
        ref: this.root
      }, Object.keys(cellSizeCache).length && _react["default"].createElement(_reactVirtualized.ScrollSync, null, function (_ref5) {
        var _onScroll = _ref5.onScroll,
            scrollLeft = _ref5.scrollLeft,
            scrollTop = _ref5.scrollTop;
        return _react["default"].createElement("div", {
          className: "results-table-wrapper"
        }, hasPinnedColumns && _react["default"].createElement("div", {
          key: "pinned-columns",
          className: "pinned-columns grid-row"
        }, _react["default"].createElement(TableSection, {
          classList: {
            header: 'pinned-columns--header pinned-grid-container',
            rows: 'pinned-columns--rows pinned-grid-container'
          },
          isPinned: true,
          columns: pinnedColumns,
          headerGridProps: headerGridProps,
          fixedWidth: pinnedColumnsWidth,
          onScroll: function onScroll(args) {
            return _onScroll(_objectSpread({}, args, {
              scrollLeft: scrollLeft
            }));
          },
          scrollTop: scrollTop,
          dataGridProps: dataGridProps,
          setGridRef: function setGridRef(pinnedGrid) {
            return _this2.pinnedGrid = pinnedGrid;
          },
          columnWidth: columnWidthFunction(pinnedColumns, cellSizeCache),
          headerCellRender: renderHeaderCell(pinnedColumns, true, _this2.props, _this2.toggleMoreOptions, moreOptionsColumn),
          dataCellRender: renderDataCell(pinnedColumns, true, _this2.props)
        })), _react["default"].createElement("div", {
          key: "unpinned-columns",
          style: {
            marginLeft: "".concat(hasPinnedColumns ? "".concat(pinnedColumnsWidth, "px") : '0')
          },
          className: "unpinned-columns grid-column"
        }, _react["default"].createElement(TableSection, {
          classList: {
            header: 'unpinned-columns--header unpinned-grid-container',
            rows: 'unpinned-columns--rows unpinned-grid-container'
          },
          isPinned: false,
          columns: unpinnedColumnsGhost,
          headerGridProps: headerGridProps,
          fixedWidth: fixedWidth,
          fixedHeight: fixedHeight,
          onScroll: _onScroll,
          scrollTop: scrollTop,
          scrollLeft: scrollLeft,
          dataGridProps: dataGridProps,
          setGridRef: function setGridRef(unpinnedGrid) {
            return _this2.unpinnedGrid = unpinnedGrid;
          },
          columnWidth: columnWidthFunction(unpinnedColumnsGhost, cellSizeCache, ghost),
          headerCellRender: renderHeaderCell(unpinnedColumnsGhost, false, _this2.props, _this2.toggleMoreOptions, moreOptionsColumn),
          dataCellRender: renderDataCell(unpinnedColumnsGhost, false, _this2.props)
        })));
      }));
    }
  }]);
  return DataTable;
}(_react.Component);

exports.DataTable = DataTable;
(0, _defineProperty2["default"])(DataTable, "defaultProps", {
  rows: [],
  pinnedColumns: [],
  colMeta: {},
  cellSizeCache: {},
  sortColumn: {},
  fixedWidth: null,
  fixedHeight: null,
  theme: {}
});

function DataTableFactory() {
  return (0, _styledComponents.withTheme)(DataTable);
}

var _default = DataTableFactory;
exports["default"] = _default;
